
<!DOCTYPE html>


<html lang="en" data-content_root="./" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Configuration &#8212; HerediClassify 1.0.6 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="_static/styles/sphinx-book-theme.css?v=a3416100" />
  
  <!-- So that users can add custom icons -->
  <script src="_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="_static/documentation_options.js?v=d728ce32"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'configuration';</script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Implemented rules" href="moduls.html" />
    <link rel="prev" title="Execution" href="execution.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search..."
         aria-label="Search..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="index.html">
  
  
  
  
  
  
    <p class="title logo__title">HerediClassify 1.0.6 documentation</p>
  
</a></div>
        <div class="sidebar-primary-item">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="execution.html">Execution</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="moduls.html">Implemented rules</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="_sources/configuration.rst" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.rst</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button>


<button class="btn btn-sm pst-navbar-icon search-button search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
</button>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Configuration</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#configuration-name">Configuration name</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#rules">Rules</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#disease-relevant-transcript">Disease relevant transcript</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#thresholds-prediction-tools">Thresholds prediction tools</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#likelihood-thresholds">Likelihood thresholds</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#allele-frequency-thresholds">Allele frequency thresholds</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#functional-thresholds">Functional thresholds</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#annotation-files">Annotation files</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-specific-configurations">Gene-specific configurations</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="configuration">
<h1>Configuration<a class="headerlink" href="#configuration" title="Link to this heading">#</a></h1>
<p>A number of different thresholds and paths can be defined in the configuration file.
An example configuration <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> is provided in the repository.
Additionally, configuration files for the gene specific recommendations for the hereditary breast and ovarian cancer risk genes <em>ATM</em> (v.1.3.0), <em>BRCA1</em> (v.1.1.0), <em>BRCA2</em> (v.1.1.0), <em>CDH1</em> (v.3.1.0), <em>PALB2</em> (v.1.1.0), <em>PTEN</em> (v.3.1.0), and <em>TP53</em> (v.1.4.0) are provided (see <code class="docutils literal notranslate"><span class="pre">gene_specific</span></code> folder).</p>
<p>For further information on the gene specific guidelines please see the documentation on <a class="reference external" href="https://www.clinicalgenome.org/affiliation/vcep/#ep_table_heading">ClinGen</a>.</p>
<section id="configuration-name">
<h2>Configuration name<a class="headerlink" href="#configuration-name" title="Link to this heading">#</a></h2>
<p>At the top at each configuration file a <code class="docutils literal notranslate"><span class="pre">name</span></code> and a <code class="docutils literal notranslate"><span class="pre">version</span></code> can be defined.
Version and name of the configuration together with the final classification result.</p>
</section>
<section id="rules">
<h2>Rules<a class="headerlink" href="#rules" title="Link to this heading">#</a></h2>
<p>In the rules section of the configuration file, a list of rules from the implemented rules can be defined in a list.
The processing of the rules is not case sensitive.</p>
</section>
<section id="disease-relevant-transcript">
<h2>Disease relevant transcript<a class="headerlink" href="#disease-relevant-transcript" title="Link to this heading">#</a></h2>
<p>In this section of the configuration file a disease relevant transcript or list of disease relevant transcripts can be defined.
If this is not the case all transcripts will be analysed and returned, with results for the MANE transcript being returned as the main result.</p>
<p>The <code class="docutils literal notranslate"><span class="pre">name</span></code> needs to be the Ensembl transcript ID without version extension.
Under <code class="docutils literal notranslate"><span class="pre">nmd_threshold</span></code> the last nucleotide position that may cause nonsense mediated decay can be documented.
<code class="docutils literal notranslate"><span class="pre">pos_last_known_patho_ptc</span></code> allows for the definition of the last known pathogenic premature termination codon.
These transcript specific information are needed for the application of <strong>PVS1</strong>, especially in the gene-specific recommendations.</p>
</section>
<section id="thresholds-prediction-tools">
<h2>Thresholds prediction tools<a class="headerlink" href="#thresholds-prediction-tools" title="Link to this heading">#</a></h2>
<p>In this section thresholds for pathogenicitiy prediction tools can be defined.
HerediClassify is only designed to use on prediction tool for pathogenicity prediction and one prediction tool for splicing prediction.
For pathogenicity prediction the thresholds can be defined under <code class="docutils literal notranslate"><span class="pre">pathogenicity_prediction</span></code> and thresholds for splicing prediction tools can be defined under <code class="docutils literal notranslate"><span class="pre">splicing_prediction</span></code>.
Both thresholds have the same structure.
Under <code class="docutils literal notranslate"><span class="pre">name</span></code> the name of the prediction tool can be set.
The prediction tool should be the same as the one specific in the variant import json.
Under <code class="docutils literal notranslate"><span class="pre">benign</span></code> and <code class="docutils literal notranslate"><span class="pre">pathogenic</span></code> the thresholds for pathogenic and benign evidence can be set.
Using <code class="docutils literal notranslate"><span class="pre">direction</span></code> the direction of the comparison between the threshold and the prediction value is defined.
The following directions are possible:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">less</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">less_than_or_equal</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">greater</span></code></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">greater_than_or_equal</span></code></p></li>
</ul>
<p>To define the thresholds use <code class="docutils literal notranslate"><span class="pre">evidence</span> <span class="pre">strength:threshold</span></code> e.g. <code class="docutils literal notranslate"><span class="pre">very_strong:0.003</span></code>.</p>
<p>The thresholds for splicing and pathogenicity predictions at least at supporting level are required for all implementations of <strong>BP4</strong> and <strong>PP3</strong>.
Furthermore, the threshold for splicing prediction is required for the application of <strong>PVS1</strong> and <strong>BP7</strong>.
Additional, the threshold for splicing prediction is required for <strong>PM5</strong>, <strong>PM5_PTEN</strong>, <strong>PM5_TP53</strong>, <strong>BP1_annotation_cold_spot_strong</strong>, <strong>PM1_TP53</strong>, <strong>PS1_protein_spliceai</strong>, <strong>PS1_protein_enigma</strong>, <strong>PS1_splicing</strong>, <strong>PS1_splicing_clingen</strong>, <strong>PS1_protein_TP53</strong>, <strong>PS1_splicing_TP53</strong>, and <strong>PS1_splicing_pten</strong>.</p>
<p>When using the rules <strong>PP3_protein_mult_strength</strong> or <strong>BP4_protein_mult_strength</strong> thresholds for higher evidence strengths can be defined using <code class="docutils literal notranslate"><span class="pre">moderate</span></code> , <code class="docutils literal notranslate"><span class="pre">strong</span></code>, and <code class="docutils literal notranslate"><span class="pre">very_strong</span></code>.
For example see example configuration files under <code class="docutils literal notranslate"><span class="pre">gene_sepcific/pejaver_mult_strength</span></code>.
When another implementation of <strong>PP3</strong> or <strong>BP4</strong> is used that does not apply an evidence strength higher than supporting, thresholds for higher evidence strengths will be ignored during the assessment of the criterion.</p>
</section>
<section id="likelihood-thresholds">
<h2>Likelihood thresholds<a class="headerlink" href="#likelihood-thresholds" title="Link to this heading">#</a></h2>
<p>Under likelihood thresholds the threshold for all evidence strengths can be set.
These thresholds are accessed by <strong>BS4</strong>, <strong>PP1</strong>, <strong>PP4_engima</strong>, and <strong>BP5_engima</strong>.</p>
</section>
<section id="allele-frequency-thresholds">
<h2>Allele frequency thresholds<a class="headerlink" href="#allele-frequency-thresholds" title="Link to this heading">#</a></h2>
<p>Here allele frequency thresholds can be defined as decimal figures.
A threshold of <code class="docutils literal notranslate"><span class="pre">10%</span></code> would therefore be set as <code class="docutils literal notranslate"><span class="pre">0.1</span></code>.</p>
<ul>
<li><p><strong>threshold_ba1</strong></p>
<p>Threshold for BA1, required for all <strong>BA1</strong> implementations.</p>
</li>
<li><p><strong>threshold_ba1_absolute</strong></p>
<p>Threshold for minimum absolute allele count in subpopulation for BA1 to apply, required for <strong>BA1_with_absolute</strong>.</p>
</li>
<li><p><strong>threshold_bs1</strong></p>
<p>Threshold for BS1, required for all <strong>BS1</strong> implementations.</p>
</li>
<li><p><strong>threshold_bs1_supporting</strong></p>
<p>Threshold for BS1 with supporting evidence, required for rule <strong>BS1_supporting</strong>.</p>
</li>
<li><p><strong>threshold_bs1_absolute</strong></p>
<p>Threshold for minimum absolute allele count in subpopulation for BS1 to apply, required for <strong>BS1_with_absolute</strong>.</p>
</li>
<li><p><strong>threshold_bs2</strong></p>
<p>Threshold for BS2, required for all <strong>BS2</strong> implementations.</p>
</li>
<li><p><strong>threshold_bs2_supporting</strong></p>
<p>Threshold for BS2 with supporting evidence, required for rule <strong>BS2_supporting</strong>.</p>
</li>
<li><p><strong>threshold_pm2</strong></p>
<p>Threshold for PM2, required for all <strong>PM2</strong> implementations.</p>
</li>
<li><p><strong>threshold_cancerhotspots_ac</strong></p>
<p>Threshold for the application of mutational hotspot criterion, required for <strong>PM1_TP53</strong>.</p>
</li>
</ul>
</section>
<section id="functional-thresholds">
<h2>Functional thresholds<a class="headerlink" href="#functional-thresholds" title="Link to this heading">#</a></h2>
<p>Under functional thresholds the threshold for difference in length of protein in percent can be defined. This us ed for the implementation of <strong>PVS</strong> and <strong>PM4</strong>. Please use a decimal point. So for a threshold of 10% <code class="docutils literal notranslate"><span class="pre">threshold_diff_len_prot_percent</span></code> should be set to 0.1.</p>
</section>
<section id="annotation-files">
<h2>Annotation files<a class="headerlink" href="#annotation-files" title="Link to this heading">#</a></h2>
<p>All the annotation files needed can be downloaded using the Download scripts provided and explained under installation instructions.</p>
</section>
<section id="gene-specific-configurations">
<h2>Gene-specific configurations<a class="headerlink" href="#gene-specific-configurations" title="Link to this heading">#</a></h2>
<p>Under this header the location of the gene-specific configuration is defined.
Under root the folder where the gene-specific configuration are located is defined.
Underneath, every gene name can be mapped to a specific file in which the gene-specific configuration is housed.
After loading the variant and the configuration, HerediClassify checks if there is a gene-specific configuration available and uses that.
When no gene-specific configuration are defined, HerediClassify will use the parameters of the general configuration file for all variants.
The definition of the gene-specific configuration is only needed in the general configuration file.</p>
</section>
</section>


                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="execution.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Execution</p>
      </div>
    </a>
    <a class="right-next"
       href="moduls.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Implemented rules</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#configuration-name">Configuration name</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#rules">Rules</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#disease-relevant-transcript">Disease relevant transcript</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#thresholds-prediction-tools">Thresholds prediction tools</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#likelihood-thresholds">Likelihood thresholds</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#allele-frequency-thresholds">Allele frequency thresholds</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#functional-thresholds">Functional thresholds</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#annotation-files">Annotation files</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-specific-configurations">Gene-specific configurations</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By Anna-Lena Katzke
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2024, Anna-Lena Katzke.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>