<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Execution &#8212; HerediClassify 1.0.6 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=27fed22d" />
    <script src="_static/documentation_options.js?v=e2a723ec"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="execution">
<h1>Execution<a class="headerlink" href="#execution" title="Link to this heading">¶</a></h1>
<p>HerediClassify can either be run in the command line, alternatively a FastAPI is implemented</p>
<section id="execution-in-the-command-line">
<h2>Execution in the command line<a class="headerlink" href="#execution-in-the-command-line" title="Link to this heading">¶</a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>variant_classification/classify.py<span class="w"> </span>-c<span class="w"> </span>config.yaml<span class="w"> </span>-p<span class="w"> </span>json_string
</pre></div>
</div>
</section>
<section id="execution-via-fastapi">
<h2>Execution via FastAPI<a class="headerlink" href="#execution-via-fastapi" title="Link to this heading">¶</a></h2>
<p><strong>1. Start the FastAPI server</strong></p>
<p>This will start a uvicorn server running on <a class="reference external" href="http://0.0.0.0:8080">http://0.0.0.0:8080</a></p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python<span class="w"> </span>webservice.py
</pre></div>
</div>
<p><strong>2. Execute classify on server</strong></p>
<p>Send a curl request using the following command to the server</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>curl<span class="w"> </span>-X<span class="w"> </span><span class="s1">&#39;POST&#39;</span><span class="w"> </span><span class="se">\</span>
<span class="s1">&#39;http://0.0.0.0:8080/classify_variant&#39;</span><span class="w"> </span><span class="se">\</span>
-H<span class="w"> </span><span class="s1">&#39;accept: application/json&#39;</span><span class="w"> </span><span class="se">\</span>
-H<span class="w"> </span><span class="s1">&#39;Content-Type: application/json&#39;</span><span class="w"> </span><span class="se">\</span>
-d<span class="w"> </span><span class="s1">&#39;{</span>
<span class="s1">&quot;config_path&quot;: &quot;/home/katzkean/variant_classification/config.yaml&quot;,</span>
<span class="s1">&quot;variant_json&quot;: &quot;{\&quot;chr\&quot;: \&quot;17\&quot;, \&quot;pos\&quot;: 43057110, \&quot;gene\&quot;: \&quot;BRCA1\&quot;, \&quot;ref\&quot;: \&quot;A\&quot;, \&quot;alt\&quot;: \&quot;C\&quot;, \&quot;variant_type\&quot;: [\&quot;missense_variant\&quot;], \&quot;variant_effect\&quot;: [{\&quot;transcript\&quot;: \&quot;ENST00000357654\&quot;, \&quot;hgvs_c\&quot;: \&quot;c.5219T&gt;G\&quot;, \&quot;hgvs_p\&quot;: \&quot;p.Val1740Gly\&quot;, \&quot;variant_type\&quot;: [\&quot;missense_variant\&quot;], \&quot;exon\&quot;: 19}, {\&quot;transcript\&quot;: \&quot;ENST00000471181\&quot;, \&quot;hgvs_c\&quot;: \&quot;c.5282T&gt;G\&quot;, \&quot;hgvs_p\&quot;: \&quot;p.Val1761Gly\&quot;, \&quot;variant_type\&quot;: [\&quot;missense_variant\&quot;], \&quot;exon\&quot;: 20}], \&quot;splicing_prediction_tools\&quot;: {\&quot;SpliceAI\&quot;: 0.5}, \&quot;pathogenicity_prediction_tools\&quot;: {\&quot;REVEL\&quot;: 0.5, \&quot;BayesDel\&quot;: 0.5}, \&quot;gnomAD\&quot;: {\&quot;AF\&quot;: 0.007, \&quot;AC\&quot;: 12, \&quot;popmax\&quot;: \&quot;EAS\&quot;, \&quot;popmax_AF\&quot;: 0.009, \&quot;popmax_AC\&quot;: 5}, \&quot;FLOSSIES\&quot;: {\&quot;AFR\&quot;: 9, \&quot;EUR\&quot;: 130}, \&quot;mRNA_analysis\&quot;: {\&quot;performed\&quot;: true, \&quot;pathogenic\&quot;: true, \&quot;benign\&quot;: true}, \&quot;functional_data\&quot;: {\&quot;performed\&quot;: true, \&quot;pathogenic\&quot;: true, \&quot;benign\&quot;: true}, \&quot;prior\&quot;: 0.25, \&quot;co-occurrence\&quot;: 0.56, \&quot;segregation\&quot;: 0.56, \&quot;multifactorial_log-likelihood\&quot;: 0.56, \&quot;VUS_task_force_domain\&quot;: true, \&quot;cancer_hotspot\&quot;: true, \&quot;cold_spot\&quot;: true}&quot;</span>
<span class="s1">}&#39;</span>
</pre></div>
</div>
<p>This will create the following output, with metadata included at the end</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="o">{</span><span class="s2">&quot;result&quot;</span>:<span class="s2">&quot;{\&quot;PVS1\&quot;: {\&quot;rule_type\&quot;: \&quot;general\&quot;, \&quot;evidence_type\&quot;: \&quot;pathogenic\&quot;, \&quot;status\&quot;: false, \&quot;strength\&quot;: \&quot;very_strong\&quot;, \&quot;comment\&quot;: \&quot;PVS1 does not apply to this variant, as PVS1 does not apply to variant types missense_variant.\&quot;}, \&quot;PS1_protein\&quot;: {\&quot;rule_type\&quot;: \&quot;protein\&quot;, \&quot;evidence_type\&quot;: \&quot;pathogenic\&quot;, \&quot;status\&quot;: false, \&quot;strength\&quot;: \&quot;strong\&quot;, \&quot;comment\&quot;: \&quot;No ClinVar entries found that show the same amino acid change as pathogneic.\&quot;}, \&quot;PS1_splicing\&quot;: {\&quot;rule_type\&quot;: \&quot;splicing\&quot;, \&quot;evidence_type\&quot;: \&quot;pathogenic\&quot;, \&quot;status\&quot;: false, \&quot;strength\&quot;: \&quot;strong\&quot;, \&quot;comment\&quot;: \&quot;No ClinVar entries found that show splice variants at the same nucleotide position as pathogenic..\&quot;}, \&quot;PM1\&quot;: {\&quot;rule_type\&quot;: \&quot;general\&quot;, \&quot;evidence_type\&quot;: \&quot;pathogenic\&quot;, \&quot;status\&quot;: true, \&quot;strength\&quot;: \&quot;moderate\&quot;, \&quot;comment\&quot;: \&quot;Variant in mutational hotspot.\&quot;}, \&quot;PM2\&quot;: {\&quot;rule_type\&quot;: \&quot;general\&quot;, \&quot;evidence_type\&quot;: \&quot;pathogenic\&quot;, \&quot;status\&quot;: false, \&quot;strength\&quot;: \&quot;moderate\&quot;, \&quot;comment\&quot;: \&quot;Variant occures with 0.009 in GnomAD subpopulation EAS.\&quot;}, \&quot;PM4\&quot;: {\&quot;rule_type\&quot;: \&quot;general\&quot;, \&quot;evidence_type\&quot;: \&quot;pathogenic\&quot;, \&quot;status\&quot;: false, \&quot;strength\&quot;: \&quot;moderate\&quot;, \&quot;comment\&quot;: \&quot;PM4 does not apply to this variant, as PVS1 does not apply to variant types missense_variant.\&quot;}, \&quot;PM5_protein\&quot;: {\&quot;rule_type\&quot;: \&quot;protein\&quot;, \&quot;evidence_type\&quot;: \&quot;pathogenic\&quot;, \&quot;status\&quot;: false, \&quot;strength\&quot;: \&quot;moderate\&quot;, \&quot;comment\&quot;: \&quot;No ClinVar entries found that show an amino acid change in the same position as pathogenic.\&quot;}, \&quot;PM5_splicing\&quot;: {\&quot;rule_type\&quot;: \&quot;splicing\&quot;, \&quot;evidence_type\&quot;: \&quot;pathogenic\&quot;, \&quot;status\&quot;: false, \&quot;strength\&quot;: \&quot;moderate\&quot;, \&quot;comment\&quot;: \&quot;No ClinVar entries found that show variant in the same splice site as pathogenic.\&quot;}, \&quot;PP3_protein\&quot;: {\&quot;rule_type\&quot;: \&quot;protein\&quot;, \&quot;evidence_type\&quot;: \&quot;pathogenic\&quot;, \&quot;status\&quot;: false, \&quot;strength\&quot;: \&quot;supporting\&quot;, \&quot;comment\&quot;: \&quot;Variant is not predicted to be pathogenic by REVEL.\&quot;}, \&quot;PP3_splicing\&quot;: {\&quot;rule_type\&quot;: \&quot;splicing\&quot;, \&quot;evidence_type\&quot;: \&quot;pathogenic\&quot;, \&quot;status\&quot;: true, \&quot;strength\&quot;: \&quot;supporting\&quot;, \&quot;comment\&quot;: \&quot;Variant is predicted to have a splice effect by SpliceAI.\&quot;}, \&quot;BA1\&quot;: {\&quot;rule_type\&quot;: \&quot;general\&quot;, \&quot;evidence_type\&quot;: \&quot;benign\&quot;, \&quot;status\&quot;: false, \&quot;strength\&quot;: \&quot;stand_alone\&quot;, \&quot;comment\&quot;: \&quot;Variant occures with 0.009 in GnomAD subpopulation EAS.\&quot;}, \&quot;BS1\&quot;: {\&quot;rule_type\&quot;: \&quot;general\&quot;, \&quot;evidence_type\&quot;: \&quot;benign\&quot;, \&quot;status\&quot;: false, \&quot;strength\&quot;: \&quot;strong\&quot;, \&quot;comment\&quot;: \&quot;Variant occures with 0.009 in GnomAD subpopulation EAS.\&quot;}, \&quot;BS2\&quot;: {\&quot;rule_type\&quot;: \&quot;general\&quot;, \&quot;evidence_type\&quot;: \&quot;benign\&quot;, \&quot;status\&quot;: true, \&quot;strength\&quot;: \&quot;strong\&quot;, \&quot;comment\&quot;: \&quot;The variant occures 130 in FLOSSIES.\&quot;}, \&quot;BP3\&quot;: {\&quot;rule_type\&quot;: \&quot;general\&quot;, \&quot;evidence_type\&quot;: \&quot;benign\&quot;, \&quot;status\&quot;: false, \&quot;strength\&quot;: \&quot;supporting\&quot;, \&quot;comment\&quot;: \&quot;BP3 does not apply to this variant, as BP3 does not apply to variant types missense_variant.\&quot;}, \&quot;BP4_protein\&quot;: {\&quot;rule_type\&quot;: \&quot;protein\&quot;, \&quot;evidence_type\&quot;: \&quot;benign\&quot;, \&quot;status\&quot;: false, \&quot;strength\&quot;: \&quot;supporting\&quot;, \&quot;comment\&quot;: \&quot;Variant is not predicted to be benign REVEL.\&quot;}, \&quot;BP4_splicing\&quot;: {\&quot;rule_type\&quot;: \&quot;splicing\&quot;, \&quot;evidence_type\&quot;: \&quot;benign\&quot;, \&quot;status\&quot;: false, \&quot;strength\&quot;: \&quot;supporting\&quot;, \&quot;comment\&quot;: \&quot;Variant is not predicted to have no splicing effect by SpliceAI.\&quot;}, \&quot;BP7_splicing\&quot;: {\&quot;rule_type\&quot;: \&quot;splicing\&quot;, \&quot;evidence_type\&quot;: \&quot;benign\&quot;, \&quot;status\&quot;: false, \&quot;strength\&quot;: \&quot;supporting\&quot;, \&quot;comment\&quot;: \&quot;Variant is not predicted to have no splicing effect by SpliceAI.\&quot;}}&quot;</span>,
<span class="s2">&quot;config_file&quot;</span>:<span class="s2">&quot;config.yaml&quot;</span>,
<span class="s2">&quot;config_name&quot;</span>:<span class="s2">&quot;acmg brca2&quot;</span>,
<span class="s2">&quot;date&quot;</span>:<span class="s2">&quot;2023-12-04&quot;</span>,
<span class="s2">&quot;version&quot;</span>:<span class="s2">&quot;0.1.0&quot;</span><span class="o">}</span>
</pre></div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">HerediClassify</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2024, Anna-Lena Katzke.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="_sources/execution.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>