<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Implemented rules &#8212; HerediClassify 1.0.6 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=d1102ebc" />
    <link rel="stylesheet" type="text/css" href="_static/basic.css?v=686e5160" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=27fed22d" />
    <script src="_static/documentation_options.js?v=e2a723ec"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Execution" href="execution.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="implemented-rules">
<h1>Implemented rules<a class="headerlink" href="#implemented-rules" title="Link to this heading">¶</a></h1>
<p>The decision trees for all implemented criteria are listed below.
Please also check the gene specific recommendations for more details.</p>
<section id="pvs1">
<h2>PVS1<a class="headerlink" href="#pvs1" title="Link to this heading">¶</a></h2>
<p>Each implementation of PVS1 implements a decision tree to determine whether or not a variant causes a loss of function for the relevant protein.</p>
<p>For further information for the general implementation of PVS1 see Abou Tayoun et al. (Human Mutation, 2018).
The gene specific implementations for PVS1 follow decision trees based on that suggested by Abou Tayoun, and specified in the corresponding gene-specific guidelines on ClinGen.</p>
<ul class="simple">
<li><p><strong>PVS1</strong></p></li>
<li><p><strong>PVS1_BRCA1</strong></p></li>
<li><p><strong>PVS1_BRCA2</strong></p></li>
<li><p><strong>PVS1_ATM</strong></p></li>
<li><p><strong>PVS1_PALB2</strong></p></li>
<li><p><strong>PVS1_PTEN</strong></p></li>
<li><p><strong>PVS1_CDH1</strong></p></li>
</ul>
</section>
<section id="ps1">
<h2>PS1<a class="headerlink" href="#ps1" title="Link to this heading">¶</a></h2>
<p>Check ClinVar for variants with the same amino acid exchange and whether or not they are (likely) pathogenic.
For specifications of the individual rules please see list of implementations below.</p>
<ul class="simple">
<li><p><strong>PS1_protein</strong>:
Check for variant with same amino acid exchange in ClinVar.
ClinVar classification has to be pathogenic or likely pathogenic/pathogenic in order for a variant to be considered for PS1 evidence.
A likely pathogenic classification on ClinVar is not sufficient.</p></li>
<li><p><strong>PS1_protein_spliceai</strong>:
Same as PS1_protein.
In addition, the variant under assessment and ClinVar variants must be predicted to not affect splicing.</p></li>
<li><p><strong>PS1_protein_enigma</strong>:
Same as PS1_protein_spliceai.
In case of only likely pathogenic variants with the same amino acid exchange, the evidence strength of PS1 is downgraded to moderate.</p></li>
<li><p><strong>PS1_splicing</strong>:
Variant at same position has been classified as pathogenic.</p></li>
<li><p><strong>PS1_splicing_clingen</strong>:
Implementing the decision tree for splicing variants, as recommended by Walker et al (AJHG, 2023).</p></li>
<li><p><strong>PS1_protein_tp53</strong>:
See recommendation for the application of PS1 for TP53 on ClinGen.</p></li>
<li><p><strong>PS1_splicing_tp53</strong>:
See recommendation for the application of PS1 for TP53 on ClinGen.</p></li>
<li><p><strong>PS1_splicing_pten</strong>:
See recommendations for the application of PS1 for PTEN on ClinGen</p></li>
</ul>
</section>
<section id="ps3">
<h2>PS3<a class="headerlink" href="#ps3" title="Link to this heading">¶</a></h2>
<p>Functional assay indicated pathogenicity of variant.</p>
<ul class="simple">
<li><p>PS3</p></li>
</ul>
</section>
<section id="pm1">
<h2>PM1<a class="headerlink" href="#pm1" title="Link to this heading">¶</a></h2>
<p>Variant located in mutational hotspot.</p>
<ul class="simple">
<li><p><strong>PM1_supporting</strong>:
PM1 is only implemented based on supporting evidence strength.
For the identification of hotspot regions, see download scripts.
In the configuration file under annotation_files/critical_region/critical_regions a bed file with defined critical regions can be defined that is being used for PM1.</p></li>
<li><p><strong>PM1_TP53</strong>:
See recommendations for the application of PM1 for TP53</p></li>
</ul>
</section>
<section id="pm2">
<h2>PM2<a class="headerlink" href="#pm2" title="Link to this heading">¶</a></h2>
<p>Variant absent from healthy population.</p>
<ul class="simple">
<li><p><strong>PM2</strong>:
Applies PM2 with moderate evidence strength.
PM2 will be using the popmax allele frequency and fall back on general allele frequency should popmax not be available.
Allele frequency has to be less than or equal to threshold set for PM2.</p></li>
<li><p><strong>PM2_supporting_faf</strong>:
Uses filtered allele frequency instead of popmax allele frequeny.</p></li>
<li><p><strong>PM2_supporting_less</strong>:
When comparing allele frequency to allele frequency, the allele frequency has to be smaller than the threshold.</p></li>
<li><p><strong>PM2_supporting_less_faf</strong>:
When comparing allele frequency to allele frequency, the allele frequency has to be smaller than the threshold.
Uses filtered allele frequency instead of popmax allele frequeny.</p></li>
<li><p><strong>PM2_supporting_no_indel</strong>:
PM2 will not be applied to indels.</p></li>
<li><p><strong>PM2_supporting_no_indel_faf</strong>:
PM2 will not be applied to indels.
Uses filtered allele frequency instead of popmax allele frequeny.</p></li>
</ul>
</section>
<section id="pm4">
<h2>PM4<a class="headerlink" href="#pm4" title="Link to this heading">¶</a></h2>
<p>Length change of protein larger than 10% and located in disease relevant region.</p>
<ul class="simple">
<li><p><strong>PM4</strong>:
PM4 applies to inframe insertions or deletions, stop loss variants, and frameshift variants.
For inframe variants, deletion/insertion has to be either located in a disease relevant region, or the change in protein length must be greater than 10%.
PM4 is applied to all stop loss variants that extend the protein by more than 10%. In case a repetitive region is affected PM4 does not apply.
PM4 applies to frameshift variants in case protein length is changed by more than 10% and no repetitive region is affected.</p></li>
<li><p><strong>PM4_stoploss</strong>:
PM4 only applies to stop loss variants that increase protein length by more than 10% and affect a disease relevant region.</p></li>
<li><p><strong>PM4_PTEN</strong>:
See recommendations for the application of PM4 for PTEN.
PM4 applies to inframe deletions and insertions if they are located in disease relevant region and to all increases in protein length if not located in repetitive regions.</p></li>
</ul>
</section>
<section id="pm5">
<h2>PM5<a class="headerlink" href="#pm5" title="Link to this heading">¶</a></h2>
<p>ClinVar entry for same amino acid position with different amino acid exchange has been shown to be pathogenic.</p>
<ul class="simple">
<li><p><strong>PM5_protein</strong>:
PM5 applies if a variant with different amino acid exchange at same position has been classifed as pathogenic, pathogenic/likely pathogenic or likely pathogenic in ClinVar.</p></li>
<li><p><strong>PM5_protein_pathogenic</strong>:
PM5 applies if a variant with different amino acid exchange at same position has been classifed as pathogenic or likely pathogenic/ pathogenic in ClinVar.
A likely pathogenic classification on ClinVar is not sufficient.</p></li>
<li><p><strong>PM5_protein_ptc</strong>:
PM5 is applicable if the premature termination codon (PTC) caused by the variant is located upstream of the last known pathogenic PTC.
This applies to variant types: frameshift, stop gained and stop loss.
The last known PTC has to be provided in the configuration file under disease_relevant_transcripts under pos_last_known_patho_ptc (int). Please also specify which transcript (Ensemble) the PTC is located in.
The last kown PTC can be defined for multiple transcripts.</p></li>
<li><p><strong>PM5_splicing_ptc</strong>:
PM5 is applicable if the premature termination codon (PTC) caused by the variant is located upstream of the last known pathogenic PTC.
This applies to variant types: splice acceptor variant and splice donor variant.
The last known PTC has to be provided in the configuration file under disease_relevant_transcripts under pos_last_known_patho_ptc (int). Please also specify which transcript (Ensemble) the PTC is located in.
The last kown PTC can be defined for multiple transcripts.</p></li>
<li><p><strong>PM5_protein_CDH1</strong>:
PM5 is applicable if the premature termination codon (PTC) caused by the variant is located upstream of the last known pathogenic PTC.
Alternatively, PM5 applies when a variant is predicted to cause nonsense mediated decay.
This applies to variant types: frameshift, stop gained and stop loss.
The last known PTC has to be provided in the configuration file under disease_relevant_transcripts under pos_last_known_patho_ptc (int). Please also specify which transcript (Ensemble) the PTC is located in.</p></li>
<li><p><strong>PM5_splicing_CDH1</strong>:
PM5 is applicable if the premature termination codon (PTC) caused by the variant is located upstream of the last known pathogenic PTC.
Alternatively, PM5 applies when a variant is predicted to cause nonsense mediated decay.
This applies to variant types: splice acceptor variant and splice donor variant.
The last known PTC has to be provided in the configuration file under disease_relevant_transcripts under pos_last_known_patho_ptc (int). Please also specify which transcript (Ensemble) the PTC is located in.</p></li>
<li><p><strong>PM5_enigma</strong>:
PM5 applies based on the exon the premature termination codon is located in.
A table with the necessary data is provided under data/PM5_annotations/PM5_PTC_BRCA1.csv</p></li>
<li><p><strong>PM5_protein_PTEN</strong>:
PM5 applies if a variant with different amino acid exchange at same position has been classified as pathogenic, pathogenic/likely pathogenic or likely pathogenic in ClinVar.
A splicing effect has to be excluded for the variant under assessment as well as for the variants in ClinVar.
Additionally, the blosum62 score of the variant under assessments needs to be smaller than that of the ClinVar variants.</p></li>
<li><p><strong>PM5_protein_TP53</strong>:
PM5 applies if a variant with different amino acid exchange at same position has been classified as pathogenic, pathogenic/likely pathogenic or likely pathogenic in ClinVar.
A splicing effect has to be excluded for the variant under assessment as well as for the variants in ClinVar.
Additionally, the Grantham score of the variant under assessments needs to be greater than that of the ClinVar variants.</p></li>
</ul>
</section>
<section id="pp1">
<h2>PP1<a class="headerlink" href="#pp1" title="Link to this heading">¶</a></h2>
<p>Variant segregates with disease.</p>
<ul class="simple">
<li><p><strong>PP1</strong>:
Variant segregates with disease.
Threshold for likelihood can be set under likelihood_thresholds for benignity and pathogenicity and differnet evidence strengths.</p></li>
</ul>
</section>
<section id="pp2">
<h2>PP2<a class="headerlink" href="#pp2" title="Link to this heading">¶</a></h2>
<p>Gene is known to have any pathogenic variants of the same type.</p>
<ul class="simple">
<li><p><strong>PP2</strong>:
PP2 applies to all missense variants.</p></li>
</ul>
</section>
<section id="pp3">
<h2>PP3<a class="headerlink" href="#pp3" title="Link to this heading">¶</a></h2>
<p>Computational evidence for pathogenicity of variant.</p>
<ul class="simple">
<li><p><strong>PP3_splicing</strong>:
Checks if variant is predicted to be pathogenic by prediction tool.
Threshold can be set under prediction_tool_thresholds/splicing_prediction/pathogenic/supporting (type:float).</p></li>
<li><p><strong>PP3_splicing_enigma</strong>:
Checks if variant is predicted to be pathogenic by prediction tool.
Threshold can be set under prediction_tool_thresholds/splicing_prediction/pathogenic/supporting (type:float).
PP3 can not be applied if variant is located outside of disease relevant regions.</p></li>
<li><p><strong>PP3_splicing_enigma_mult_strength</strong>:
PP3 can not be applied if variant is located outside of disease relevant regions.
Thresholds to be used for PP3 for all evidence strength levels are assessed and can be set under prediction_tool_thresholds/splicing_prediction/pathogenic/(supporting or moderate or strong or very_strong) (type:float).</p></li>
<li><p><strong>PP3_splicing_mult_strength</strong>:
Thresholds to be used for PP3 for all evidence strength levels are assessed and can be set under prediction_tool_thresholds/splicing_prediction/pathogenic/(supporting or moderate or strong or very_strong) (type:float).</p></li>
<li><p><strong>PP3_splicing_CDH1</strong>:
PP3 only applies to intronic variants located outside of the canonical splice site.
Checks if variant is predicted to be pathogenic by prediction tool.
Threshold can be set under prediction_tool_thresholds/splicing_prediction/pathogenic/supporting (type:float).</p></li>
<li><p><strong>PP3_protein</strong>:
Checks if variant is predicted to be pathogenic by prediction tool.
Threshold can be set under prediction_tool_thresholds/pathogenicity_prediction/pathogenic/supporting (type:float).</p></li>
<li><p><strong>PP3_protein_enigma</strong>:
Checks if variant is predicted to be pathogenic by prediction tool.
Threshold can be set under prediction_tool_thresholds/pathogenicity_prediction/pathogenic/supporting (type:float).
PP3 can not be applied if variant is located outside of disease relevant regions.</p></li>
<li><p><strong>PP3_protein_enigma_mult_strength</strong>:
Thresholds to be used for PP3 for all evidence strength levels are assessed and can be set under prediction_tool_thresholds/pathogenicity_prediction/pathogenic/(supporting or moderate or strong or very_strong) (type:float).
PP3 can not be applied if variant is located outside of disease relevant regions.</p></li>
<li><p><strong>PP3_protein_mult_strength</strong>:
Thresholds to be used for PP3 for all evidence strength levels are assessed and can be set under prediction_tool_thresholds/pathogenicity_prediction/pathogenic/(supporting or moderate or strong or very_strong) (type:float).</p></li>
</ul>
</section>
<section id="ba1">
<h2>BA1<a class="headerlink" href="#ba1" title="Link to this heading">¶</a></h2>
<p>Variant is very common in helathy popualtion.</p>
<ul class="simple">
<li><p><strong>BA1</strong>:
BA1 is using the popmax allele frequency and falls back on genereal allele frequency should popmax not be available.
Threshold can be set under allele_frequency_thresholds/threshold_ba1.</p></li>
<li><p><strong>BA1_faf</strong>:
BA1 is using the filtered allele frequency and falls back on genereal allele frequency should filtered allele frequency not be available.
Threshold can be set under allele_frequency_thresholds/threshold_ba1.</p></li>
<li><p><strong>BA1_with_absolute</strong>:
BA1 is using the popmax allele frequency and falls back on genereal allele frequency should popmax not be available.
Threshold can be set under allele_frequency_thresholds/threshold_ba1.
Additionally, absolute allele count in popmax allele count is checked.
Threshold can be set under allele_frequency_thresholds/threshold_ba1_absolute.</p></li>
</ul>
</section>
<section id="bs1">
<h2>BS1<a class="headerlink" href="#bs1" title="Link to this heading">¶</a></h2>
<p>Variant is common in helathy popualtion.</p>
<ul class="simple">
<li><p><strong>BS1</strong>:
BS1 is using the popmax allele frequency and falls back on general allele frequency should popmax not be available.
Threshold can be set under allele_frequency_thresholds/threshold_bs1 (type:float).</p></li>
<li><p><strong>BS1_faf</strong>:
BS1 is using the filtered allele frequency and falls back on general allele frequency should filtered allele frequency not be available.
Threshold can be set under allele_frequency_thresholds/threshold_bs1 (type:float).</p></li>
<li><p><strong>BS1_with_absolute</strong>:
BS1 is using the popmax allele frequency and falls back on general allele frequency should popmax not be available.
Threshold can be set under allele_frequency_thresholds/threshold_bs1 (type:float).
Additionally, absolute allele count in popmax allele count is checked.
Threshold can be set under allele_frequency_thresholds/threshold_bs1_absolute (type:int).</p></li>
<li><p><strong>BS1_supporting</strong>:
BS1 is using the popmax allele frequency and falls back on general allele frequency should popmax not be available.
Threshold can be set under allele_frequency_thresholds/threshold_bs1 (type:float) and allele_frequency_thresholds/threshold_bs1_supporting (type:float).</p></li>
<li><p><strong>BS1_supporting_faf</strong>:
BS1 is using the filtered allele frequency and falls back on general allele frequency should filtered allele frequency not be available.
Option to apply BS1 with strong and supporting evidence strength.
Threshold can be set under allele_frequency_thresholds/threshold_bs1 (type:float) and allele_frequency_thresholds/threshold_bs1_supporting (type:float).
Both thresholds need to be given.</p></li>
</ul>
</section>
<section id="bs2">
<h2>BS2<a class="headerlink" href="#bs2" title="Link to this heading">¶</a></h2>
<p>Mutation found in healthy individual.</p>
<ul class="simple">
<li><p><strong>BS2</strong>:
Checks FLOSSIES database for presence of this variant.
Threshold can be set under allele_frequency_thresholds/threshold_bs2 (type:int).</p></li>
<li><p><strong>BS2_supporting</strong>:
Option to apply BS2 with strong and supporting evidence strength.
Threshold can be set under allele_frequency_thresholds/threshold_bs2 (type:int) and allele_frequency_thresholds/threshold_bs2_supporting.
Both thresholds need to be given.</p></li>
</ul>
</section>
<section id="bs3">
<h2>BS3<a class="headerlink" href="#bs3" title="Link to this heading">¶</a></h2>
<p>Functional data indicating benignity.</p>
<ul class="simple">
<li><p><strong>BS3</strong></p></li>
</ul>
</section>
<section id="bs4">
<h2>BS4<a class="headerlink" href="#bs4" title="Link to this heading">¶</a></h2>
<p>Variant does not segregate with disease.</p>
<ul class="simple">
<li><p><strong>BS4</strong>:
Variant does not segregate with disease.
Threshold for likelihood can be set under likelihood_thresholds for benignity and pathogenicity and differnet evidence strengths.</p></li>
</ul>
</section>
<section id="bp1">
<h2>BP1<a class="headerlink" href="#bp1" title="Link to this heading">¶</a></h2>
<p>Missense variant in a gene where missense variants are known not to be disease causative.</p>
<ul class="simple">
<li><p><strong>BP1</strong>:
BP1 applies to all missense variants.</p></li>
<li><p><strong>BP1_annotation_cold_spot_strong</strong>:
Variant located in known cold spot region.
BP1 is applied with strong evidence strength.
Bed file with cold-spot regions can be defined under annotation_files/critical_regions/coldspot_region.</p></li>
</ul>
</section>
<section id="bp3">
<h2>BP3<a class="headerlink" href="#bp3" title="Link to this heading">¶</a></h2>
<p>Variant located in repetitive region.</p>
<ul class="simple">
<li><p><strong>BP3</strong>:
Check if variant causes differen in portein length and if length change is located in repetitive region, BP3 is applied.</p></li>
</ul>
</section>
<section id="bp4">
<h2>BP4<a class="headerlink" href="#bp4" title="Link to this heading">¶</a></h2>
<p>Computational evidence for benignity of variant.</p>
<ul class="simple">
<li><p><strong>BP4_splicing</strong>:
Checks if variant is predicted to be benign by prediction tool.
Threshold can be set under prediction_tool_thresholds/splicing_prediction/benign/supporting (type:float).</p></li>
<li><p><strong>BP4_splicing_enigma</strong>:
Checks if variant is predicted to be benign by prediction tool.
Threshold can be set under prediction_tool_thresholds/splicing_prediction/benign/supporting (type:float).
BP4 can not be applied if variant is located outside of disease relevant regions.</p></li>
<li><p><strong>BP4_splicing_enigma_mult_strength</strong>:
BP4 can not be applied if variant is located outside of disease relevant regions.
Thresholds to be used for BP4 for all evidence strength levels are assessed and can be set under prediction_tool_thresholds/splicing_prediction/benign/(supporting or moderate or strong or very_strong) (type:float).</p></li>
<li><p><strong>BP4_splicing_mult_strength</strong>:
Thresholds to be used for BP4 for all evidence strength levels are assessed and can be set under prediction_tool_thresholds/splicing_prediction/benign/(supporting or moderate or strong or very_strong) (type:float).</p></li>
<li><p><strong>BP4_protein</strong>:
Checks if variant is predicted to be benign by prediction tool.
Threshold can be set under prediction_tool_thresholds/pathogenicity_prediction/benign/supporting (type:float).</p></li>
<li><p><strong>BP4_protein_enimga</strong>:
Checks if variant is predicted to be benign by prediction tool.
Threshold can be set under prediction_tool_thresholds/pathogenicity_prediction/benign/supporting (type:float).
BP4 can not be applied if variant is located outside of disease relevant regions.</p></li>
<li><p><strong>BP4_protein_enimga_mult_strength</strong>:
Thresholds to be used for BP4 for all evidence strength levels are assessed and can be set under prediction_tool_thresholds/pathogenicity_prediction/benign/(supporting or moderate or strong or very_strong) (type:float).
BP4 can not be applied if variant is located outside of disease relevant regions.</p></li>
<li><p><strong>BP4_protein_mult_strength</strong>:
Thresholds to be used for BP4 for all evidence strength levels are assessed and can be set under prediction_tool_thresholds/pathogenicity_prediction/benign/(supporting or moderate or strong or very_strong) (type:float).</p></li>
</ul>
</section>
<section id="bp5">
<h2>BP5<a class="headerlink" href="#bp5" title="Link to this heading">¶</a></h2>
<p>Attention: BP5 in its original ACMG definition is not implemented.
Only the adaptation made in the gene-specific recommendations for BRCA1 and BRCA2 is implemented.</p>
<ul class="simple">
<li><p><strong>BP5_enigma</strong>:
Can be applied in case multifactorial likelihood analysis data is available for variant.
Threshold for likelihood can be set under likelihood_thresholds for benignity and pathogenicity and differnet evidence strengths.</p></li>
</ul>
</section>
<section id="bp7">
<h2>BP7<a class="headerlink" href="#bp7" title="Link to this heading">¶</a></h2>
<p>Check if deep intronic or synonymous variant does not affect splicing through a splicing prediction tool.
All implementation of BP7 check splice assay for benignity evidence.
If splice assay data is available results from the splicing assay are returned and prediction is not being assessed.</p>
<ul class="simple">
<li><p><strong>BP7</strong>:
Applies only to synonymous variants.</p></li>
<li><p><strong>BP7_deep_intronic_ATM</strong>:
Applies to synonymous variants and deep intronic variants located at position &gt;7 or &lt;-40.</p></li>
<li><p><strong>BP7_deep_intronic_enigma</strong>:
Applies to synonymous variants and deep intronic variants located at position &gt;=7 or &lt;=-21.</p></li>
<li><p><strong>BP7_deep_intronic_enigma_check_disease_region</strong>:
Applies to synonymous variants and deep intronic variants located at position &gt;=7 or &lt;=-21.
Checks that variant is located outside of coldspot region.
Bed file with coldspot regions can be defined under annotation_files/critical_regions/coldspot_region.</p></li>
<li><p><strong>BP7_deep_intronic_PALB2</strong>:
Applies to synonymous variants and deep intronic variants located at position &gt;7 or &lt;-21.</p></li>
</ul>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="Main">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">HerediClassify</a></h1>









<search id="searchbox" style="display: none" role="search">
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false" placeholder="Search"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script><h3>Navigation</h3>
<p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="execution.html">Execution</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Implemented rules</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#pvs1">PVS1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ps1">PS1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ps3">PS3</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pm1">PM1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pm2">PM2</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pm4">PM4</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pm5">PM5</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pp1">PP1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pp2">PP2</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pp3">PP3</a></li>
<li class="toctree-l2"><a class="reference internal" href="#ba1">BA1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#bs1">BS1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#bs2">BS2</a></li>
<li class="toctree-l2"><a class="reference internal" href="#bs3">BS3</a></li>
<li class="toctree-l2"><a class="reference internal" href="#bs4">BS4</a></li>
<li class="toctree-l2"><a class="reference internal" href="#bp1">BP1</a></li>
<li class="toctree-l2"><a class="reference internal" href="#bp3">BP3</a></li>
<li class="toctree-l2"><a class="reference internal" href="#bp4">BP4</a></li>
<li class="toctree-l2"><a class="reference internal" href="#bp5">BP5</a></li>
<li class="toctree-l2"><a class="reference internal" href="#bp7">BP7</a></li>
</ul>
</li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="execution.html" title="previous chapter">Execution</a></li>
  </ul></li>
</ul>
</div>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2024, Anna-Lena Katzke.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 8.1.3</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 1.0.0</a>
      
      |
      <a href="_sources/moduls.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>